subsys/emds/emds_flash.c

```diff
@@ -0,0 +1,434 @@
+
+	if (rc) {
+		return rc;
+	}
+
+	fs->ate_wra -= fs->ate_size;
+	return 0;
+}
+
+static int data_wrt(struct emds_fs *fs, const void *data, size_t len)
+{
+	const uint8_t *data8 = (const uint8_t *)data;
+	int rc = 0;
+	off_t offset;
+	size_t blen;
+	size_t temp_len = len;
+	uint8_t buf[EMDS_FLASH_BLOCK_SIZE];
+
+	if (!temp_len) {
+		/* Nothing to write, avoid changing the flash protection */
+		return 0;
+	}
+
+	offset = fs->offset;
+	offset += fs->data_wra_offset & ADDR_OFFS_MASK;
+
+	blen = temp_len & ~(fs->flash_params->write_block_size - 1U);
+	/* Writes multiples of 4 bytes to flash */
+	if (blen > 0) {
+		rc = flash_write(fs->flash_dev, offset, data8, blen);
+		if (rc) {
+			return rc;
+		}
+
+		temp_len -= blen;
+		offset += blen;
+		data8 += blen;
+	}
+
+	if (temp_len) {
+		(void)memcpy(buf, data8, temp_len);
+		(void)memset(buf + temp_len, fs->flash_params->erase_value,
+			     fs->flash_params->write_block_size - temp_len);
+		rc = flash_write(fs->flash_dev, offset, buf, fs->flash_params->write_block_size);
+		if (rc) {
+			return rc;
+		}
+	}
+
+	fs->data_wra_offset += align_size(fs, len);
+	return 0;
+}
+
+static int check_erased(struct emds_fs *fs, uint32_t addr, size_t len)
+{
+	size_t bytes_to_cmp;
+	uint8_t cmp[EMDS_FLASH_BLOCK_SIZE];
+	uint8_t buf[EMDS_FLASH_BLOCK_SIZE];
+
+	(void)memset(cmp, 0xff, EMDS_FLASH_BLOCK_SIZE);
+	while (len) {
+		bytes_to_cmp = MIN(EMDS_FLASH_BLOCK_SIZE, len);
+		if (flash_read(fs->flash_dev, addr, buf, bytes_to_cmp)) {
+			return -EIO;
+		}
+
+		if (memcmp(cmp, buf, bytes_to_cmp)) {
+			return -EINVAL;
+		}
+
+		len -= bytes_to_cmp;
+		addr += bytes_to_cmp;
+	}
+	return 0;
+}
+
+static int is_ate_valid(const struct emds_ate *entry)
+{
+	return entry->crc8 == crc8_ccitt(0xff, entry, offsetof(struct emds_ate, crc8));
+}
+
+static int entry_wrt(struct emds_fs *fs, uint16_t id, const void *data, size_t len)
+{
+	int rc;
+	struct emds_ate entry;
+
+	entry.id = id;
+	entry.offset = fs->data_wra_offset;
+	entry.len = (uint16_t)len;
+	entry.crc8_data = crc8_ccitt(0xff, data, len);
+	entry.crc8 = crc8_ccitt(0xff, &entry, offsetof(struct emds_ate, crc8));
+	rc = data_wrt(fs, data, len);
+	if (rc) {
+		return rc;
+	}
+
+	rc = ate_wrt(fs, &entry);
+	if (rc) {
+		return rc;
+	}
+
+	return 0;
+}
+
+static enum ate_type ate_check(struct emds_fs *fs, uint32_t addr, struct emds_ate *entry)
+{
+	uint8_t cmp_buf[fs->ate_size];
+	int rc = flash_read(fs->flash_dev, addr, entry, fs->ate_size);
+
+	if (rc) {
+		return ATE_TYPE_UNKNOWN;
+	}
+
+	memset(cmp_buf, 0, sizeof(cmp_buf));
+	if (!memcmp(entry, cmp_buf, fs->ate_size)) {
+		return ATE_TYPE_INVALIDATED;
+	}
+
+	memset(cmp_buf, fs->flash_params->erase_value, sizeof(cmp_buf));
+	if (!memcmp(entry, cmp_buf, fs->ate_size)) {
+		return ATE_TYPE_ERASED;
+	}
+
+	if (is_ate_valid(entry)) {
+		return ATE_TYPE_VALID;
+	}
+
+	return ATE_TYPE_UNKNOWN;
+}
+
+static int ate_last_recover(struct emds_fs *fs)
```

Comment:
Is ATE always last? Or data and ate entries are interleaved?

Reply:
ATE is always last.

Reply:
Is the arrangement in flash like this: <entry1><ate1><entry2><ate2> and so on?

Reply:
No, the entries is in the beginning of the flash area and the ates are in the end of the flash area.